python-checks/src/main/java/org/sonar/python/checks/DjangoRenderContextCheck.java

```diff
@@ -0,0 +1,107 @@
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6556")
+public class DjangoRenderContextCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Use an explicit context instead of passing \"locals()\" to this Django \"render\" call.";
+  private static final String SECONDARY_MESSAGE = "locals() is assigned to \"%s\" here.";
+  private static final String RENDER_FUNCTION = "django.shortcuts.render";
+  private static final String LOCALS = "locals";
+  private static final String CONTEXT_KEYWORD = "context";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, DjangoRenderContextCheck::checkForDjangoRender);
+  }
+
+  private static void checkForDjangoRender(SubscriptionContext ctx) {
+    CallExpression expression = ((CallExpression) ctx.syntaxNode());
+    Symbol symbol = expression.calleeSymbol();
+    if (symbol != null && RENDER_FUNCTION.equals(symbol.fullyQualifiedName())) {
+      checkForContextArgument(ctx, expression);
+    }
+  }
+
+  private static void checkForContextArgument(SubscriptionContext ctx, CallExpression expression) {
+    RegularArgument contextArg = TreeUtils.nthArgumentOrKeyword(2, CONTEXT_KEYWORD, expression.arguments());
+    if (contextArg != null) {
+      if (contextArg.expression().is(Tree.Kind.CALL_EXPR)) {
+        raiseIssueLocalsCallDirectly(ctx, contextArg);
+      } else if (contextArg.expression().is(Tree.Kind.NAME)) {
+        checkIfLocalsIsAssignedToContextParameter(ctx, contextArg);
+      }
+    }
+  }
+
+  private static void raiseIssueLocalsCallDirectly(SubscriptionContext ctx, RegularArgument contextArg) {
+    CallExpression maybeLocalsCall = (CallExpression) contextArg.expression();
+    Symbol localsSymbol = maybeLocalsCall.calleeSymbol();
+    if (localsSymbol != null && LOCALS.equals(localsSymbol.fullyQualifiedName())) {
+      ctx.addIssue(maybeLocalsCall, MESSAGE);
+    }
+  }
+
+  private static void checkIfLocalsIsAssignedToContextParameter(SubscriptionContext ctx, RegularArgument contextArg) {
+    Name maybeLocalsCall = (Name) contextArg.expression();
+    Symbol contextSymbol = maybeLocalsCall.symbol();
+    if (contextSymbol != null) {
+      List<Tree> assignmentStmts = contextSymbol.usages().stream()
+        .filter(usage -> usage.kind() == Usage.Kind.ASSIGNMENT_LHS)
+        .map(Usage::tree)
+        .map(usage -> TreeUtils.firstAncestorOfKind(usage, Tree.Kind.ASSIGNMENT_STMT))
+        .collect(Collectors.toList());
+      if (assignmentStmts.size() == 1) {
+        raiseIssueLocalsIsAssigned(ctx, contextArg, assignmentStmts);
+      }
+    }
+  }
+
+  private static void raiseIssueLocalsIsAssigned(SubscriptionContext ctx, RegularArgument contextArg, List<Tree> assignmentStmts) {
```

Comment:
I guess it is possible to make this part recursive just to support multiple assignments chain like:
```python
a = locals()
b = a
c = b
return render(request, "my_template.html", content_type=None, context=c) # Issue 
```
Or add a test for this and mark it as FN

Reply:
I added the recursive call but with a maximum amount of reassignment of 5. I don't it make that much sense to go any further up the tree.